aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp

```diff
@@ -1292,5 +1294,161 @@ Tensor _sparse_csr_prod_cpu(const Tensor& input, IntArrayRef dims_to_reduce, boo
+    const Tensor& self_,
+    const Tensor& other,
+    const c10::string_view reduce,
+    const c10::optional<Tensor>& row_indices_opt,
+    const c10::optional<Tensor>& ccol_indices_opt,
+    const c10::optional<Tensor>& csr2csc_opt) {
+
+  // See [Note: hacky wrapper removal for optional tensor]
+  c10::MaybeOwned<Tensor> row_indices_maybe_owned = at::borrow_from_optional_tensor(row_indices_opt);
+  const Tensor& row_indices = *row_indices_maybe_owned;
+  const Tensor& ccol_indices = c10::value_or_else(ccol_indices_opt, [] {return Tensor();});
+  const Tensor& csr2csc = c10::value_or_else(csr2csc_opt, [] {return Tensor();});
+
+  auto self = self_;
+  auto layout = self_.layout();
+  TORCH_CHECK(layout == kSparseCsr || layout == kSparseCsc,
+      "sparse_mm_reduce: expect self to be SparseCsr or SparseCsc.");
+  if (layout == kSparseCsc) {
+    self = self_.transpose(-1, -2);
+  }
+
+  sparse::impl::check_sparse_mm_reduce_impl_inputs</*train*/false>(
+      self, Tensor(), other, row_indices, ccol_indices, csr2csc);
+
+  auto op = get_reduction_enum(reduce);
+  TORCH_CHECK(op != ReductionType::PROD, "sparse_mm_reduce: reduce type of prod has not been enabled.")
+
+  auto crow = self.crow_indices();
+  auto col = self.col_indices();
+  auto val = self.values();
+
+  TORCH_CHECK(val.ndimension() == 1, "sparse_mm_reduce: expect self.values() to be 1-dimensional.")
+
+  // init output to be all zeros, for `rows` that has no nonzero elements,
+  // the corresponding rows in the output will be zero.
+  auto out = at::zeros({self.size(0), other.size(1)}, other.options());
+  auto arg_out = at::empty({0}, col.options());
+
+  int64_t nnz = self._nnz();
+  if (nnz == 0) {
+    return std::make_tuple(out, arg_out);
+  }
+
+  // only need to calculate the out args
+  // for reduce type "amax" and "amin" for training
+  bool need_arg_out = at::GradMode::is_enabled()
+      && (self.requires_grad() || other.requires_grad())
+      && (op == ReductionType::MAX || op == ReductionType::MIN);
+
+  if (!need_arg_out) {
+    spmm_reduce_stub(kCPU, out, crow, col, val, other, op);
+  } else {
+    // allocate memory and init with invalid index
+    arg_out.resize_(out.sizes());
+    arg_out.fill_(nnz);
+    spmm_reduce_arg_stub(kCPU, out, arg_out, crow, col, val, other, op);
+  }
+
+  return std::make_tuple(std::move(out), std::move(arg_out));
+}
+
+std::tuple<Tensor, Tensor> _sparse_mm_reduce_impl_backward_sparse_csr_cpu(
+    const Tensor& self_,
+    const Tensor& grad_out,
+    const Tensor& other,
+    const c10::string_view reduce,
+    const Tensor& arg_out,
+    const Tensor& row_indices,
+    const Tensor& ccol_indices,
+    const Tensor& csr2csc,
+    std::array<bool, 2> output_mask) {
+
+  auto self = self_;
+  auto layout = self_.layout();
+  TORCH_CHECK(layout == kSparseCsr || layout == kSparseCsc,
+      "sparse_mm_reduce: expect self to be SparseCsr or SparseCsc.");
+  if (layout == kSparseCsc) {
+    self = self_.transpose(-1, -2);
+  }
```

Comment:
We should probably check that
* batch dimensions are not supported
* only matrices are expected, no hybrid tensors.

Reply:
oh thanks for the remainder! updated!